hcp/src/main/scala/com/avast/clients/storage/hcp/HcpRestStorageBackend.scala

```diff
@@ -0,0 +1,196 @@
+    logger.debug(s"Checking presence of file $sha256 in HCP")
+
+    val finalUri: Uri = splitFileName(sha256).foldLeft(rootUri)(_ / _)
+
+    try {
+      val request = Request[F](
+        Method.HEAD,
+        finalUri
+      )
+
+      httpClient.fetch(request) { resp =>
+        resp.status match {
+          case Status.Ok =>
+            `Content-Length`.from(resp.headers) match {
+              case Some(`Content-Length`(length)) => F.pure(Right(HeadResult.Exists(length)))
+              case None =>
+                resp.bodyAsText.compile.last.map { body =>
+                  Left(StorageException.InvalidResponseException(resp.status.code, body.toString, "Missing Content-Length header"))
+                }
+            }
+          case Status.NotFound =>
+            F.pure(Right(HeadResult.NotFound))
+
+          case _ =>
+            resp.bodyAsText.compile.last.map { body =>
+              Left(StorageException.InvalidResponseException(resp.status.code, body.toString, "Unexpected status"))
+            }
+        }
+      }
+    } catch {
+      case NonFatal(e) => F.raiseError(e)
+    }
+  }
+
+  override def get(sha256: Sha256, dest: File): F[Either[StorageException, GetResult]] = {
+    logger.debug(s"Getting file $sha256 from HCP")
+
+    val finalUri: Uri = splitFileName(sha256).foldLeft(rootUri)(_ / _)
+
+    try {
+      val request = Request[F](
+        Method.GET,
+        finalUri
+      )
+
+      httpClient.fetch(request) { resp =>
+        resp.status match {
+          case Status.Ok => receiveStreamedFile(sha256, dest, resp)
+          case Status.NotFound => F.pure(Right(GetResult.NotFound))
+
+          case _ =>
+            resp.bodyAsText.compile.last.map { body =>
+              Left(StorageException.InvalidResponseException(resp.status.code, body.toString, "Unexpected status"))
+            }
+        }
+
+      }
+    } catch {
+      case NonFatal(e) => F.raiseError(e)
+    }
+  }
+
+  override def close(): Unit = httpClient.shutdownNow()
+
+  private def receiveStreamedFile(sha256: Sha256, dest: File, resp: Response[F]): F[Either[StorageException, GetResult]] = {
+    `Content-Length`.from(resp.headers) match {
+      case Some(clh) =>
+        val fileCopier = new FileCopier
+        val fileOs = dest.newOutputStream
+
+        resp.body.chunks
+          .map { bytes =>
+            val bis = new ByteArrayInputStream(bytes.toArray)
+            val copied = fileCopier.copy(bis, fileOs)
+            bis.close()
+            copied
+          }
+          .compile
+          .toVector
+          .map { chunksSizes =>
+            val transferred = chunksSizes.sum
+
+            fileOs.close() // all data has been transferred
+
+            if (clh.length != transferred) {
+              Left(
+                StorageException.InvalidResponseException(resp.status.code, "-stream-", s"Expected ${clh.length} B but got $transferred B"))
```

Comment:
I would like this to not be generic InvalidResponseException, because it most probably indicates corrupted data source (after few retries).